### DISCUSSION TITLE
Colour Swapping on Adafruit 3.5" TFT LCD HX8357D

### DISCUSSION THREAD

**Discussion by allanyates** [Bodmer/TFT_eSPI#1703] [Compatible displays and setup files]
I have an Adafruit 3.5" TFT LCD, based on the HX8357D driver. It is running nicely with TFT_eSPI over an SPI interface on an ESP32. I have the library set to the HXD8357D driver.

My only catch is trying to get the colours correct. I noted the suggestions regard invertDisplay and setSwapBytes, but SwapBytes did nothing, and invertDisplay gave a worse result. See attached images.

Wondering if anybody had come up with the correct settings for this display?

Thanks,

Allan
![IMG_6297_tn](https://user-images.githubusercontent.com/70734610/157969272-3c24e409-c407-40fd-b56e-7ad1c8b5cf95.JPG)

![IMG_6299_tn](https://user-images.githubusercontent.com/70734610/157969380-4bb6504a-4b32-4d44-9eda-2515f99b6104.JPG)

**Comment by Bodmer** [comment]
Which version of TFT_eSPI are you running?

**Comment by Bodmer** [comment]
It appears some TFT displays have the red and blue pixel elements swapped. This is rather odd because the default colour order is the same as the Adafruit libraries for that display.
Anyway, I have updated the master library driver files for the HX8357D display to permit the colour order to be defined in the setup file. By default the colour order is RGB (as in Adafruit libraries) but this can be changed to BGR.  The Setup_15 header has been updated with the following two lines:

// If green and blue swapped on display then change the RGB colour order
// Only uncomment ONE of the following options
#define TFT_RGB_ORDER TFT_RGB    //   Red-Green-Blue
//#define TFT_RGB_ORDER TFT_BGR    //   Blue-Green-RED

**Comment by allanyates** [comment]
I was running 2.4.42 of TFT_eSPI.

I overwrote the installed library with your latest for Github. The colours are now correct, without even tweaking either of those RGB colour order settings. Success!

I also observed that the Rotation requirement is different. Where I had to use 5 before, to avoid backwards text, the setting of 3 now gives me the same (proper result).

Many thanks for all your great efforts!

![IMG_6325_tn](https://user-images.githubusercontent.com/70734610/158045990-a31644ab-9e97-4929-893a-9d23253280f8.JPG)
